<!DOCTYPE html>

<html>

<head>
<title>canu analysis for assembly 'test' in directory '/home/deren/Documents/genomics-course/session-11/notebooks/canu-test/unitigging'</title>
<style type='text/css'>
body       { font-family: Helvetica, Verdana, sans-serif; }
h1, h2, h3 { color: #ee3e80; }
p          { color: #665544; }
th, td     { border: 1px solid #111111; padding: 2px 2px 2px 2px; }
td:hover   { background-color: #e4e4e4; }
th:hover   { background-color: #d4d4d4; }
tr.details { visibility: collapse; }
</style>
</head>

<body>

<h1>Assembly</h1>
<h2>Input Reads</h2>

<table>
<tr id='gkpload0'><td colspan='2'>/home/deren/Documents/genomics-course/session-11/notebooks/canu-test/test.trimmedReads.fasta.gz</td></tr>
<tr class='details'><td rowspan='8'>Parameters</td><td>preset = N/A</td></tr>
<tr class='details'><td>defaultQV = 20</td></tr>
<tr class='details'><td>isNonRandom = false</td></tr>
<tr class='details'><td>removeDuplicateReads = true</td></tr>
<tr class='details'><td>finalTrim = true</td></tr>
<tr class='details'><td>removeSpurReads = true</td></tr>
<tr class='details'><td>removeChimericReads = true</td></tr>
<tr class='details'><td>checkForSubReads = true</td></tr>
<tr class='details'><td rowspan='2'>FASTA</td><td>110785 reads (730952224 bp)</td></tr>
<tr class='details'><td>0 reads (0 bp) were short and not loaded</td></tr>
<tr class='details'><td rowspan='2'>FASTQ</td><td>0 reads (0 bp)</td></tr>
<tr class='details'><td>0 reads (0 bp) were short and not loaded</td></tr>
<tr><td colspan='2'>110785 reads (730952224 bp) loaded, 0 reads (0 bp) skipped, 0 warnings</td></tr>
</table>

<h2>Final Store</h2>

<table>
<tr><td colspan='2'>/home/deren/Documents/genomics-course/session-11/notebooks/canu-test/unitigging/test.gkpStore</td></tr>
<tr><td>readsLoaded</td><td>110785 reads (730952224 bp)</td></tr>
<tr><td>readsSkipped</td><td>0 reads (0 bp) (read was too short)</td></tr>
<tr><td>warnings</td><td>0 warnings (invalid base or quality value)</td></tr>
</table>
<h3>Read Length Histogram</h3>
<figure>
<a href='/home/deren/Documents/genomics-course/session-11/notebooks/canu-test/unitigging.html.files/readlengths.lg.png'><img src='/home/deren/Documents/genomics-course/session-11/notebooks/canu-test/unitigging.html.files/readlengths.sm.png'></a>
<figcaption>

</figcaption>
</figure>
<h2>k-Mer Counts</h2>

<figure>
<a href='/home/deren/Documents/genomics-course/session-11/notebooks/canu-test/unitigging.html.files/test.ms22.histogram.lg.png'><img src='/home/deren/Documents/genomics-course/session-11/notebooks/canu-test/unitigging.html.files/test.ms22.histogram.sm.png'></a>
<figcaption>
Histogram for k=22 with 728625739 mers, 43529115 distinct mers and 18857695 single-copy mers.  Largest count is 7943.
</figcaption>
</figure>
<h2>Overlaps</h2>

<table>
<tr><th>Category</th><th>Reads</th><th>%</th><th colspan='3'>Read Length</th><th colspan='3'>Feature Size or Coverage</th><th>Analysis</th></tr>
<tr><td>middle-missing        </td><td>13</td><td>0.01</td><td align='right'>11010.23</td><td>&plusmn;</td><td align='left'>8513.50</td><td align='right'>305.31</td><td>&plusmn;</td><td align='left'>392.29</td><td align='left'>bad trimming</td></tr>
<tr><td>middle-hump            </td><td>1</td><td>0.00</td><td align='right'>2399.00</td><td>&plusmn;</td><td align='left'>0.00</td><td align='right'>51.00</td><td>&plusmn;</td><td align='left'>0.00</td><td align='left'>bad trimming</td></tr>
<tr><td>no-5-prime            </td><td>44</td><td>0.04</td><td align='right'>6936.00</td><td>&plusmn;</td><td align='left'>4970.18</td><td align='right'>55.57</td><td>&plusmn;</td><td align='left'>171.86</td><td align='left'>bad trimming</td></tr>
<tr><td>no-3-prime            </td><td>29</td><td>0.03</td><td align='right'>7103.03</td><td>&plusmn;</td><td align='left'>4947.97</td><td align='right'>400.79</td><td>&plusmn;</td><td align='left'>989.61</td><td align='left'>bad trimming</td></tr>
<tr><td>low-coverage         </td><td>300</td><td>0.27</td><td align='right'>3255.88</td><td>&plusmn;</td><td align='left'>2825.85</td><td align='right'>3.91</td><td>&plusmn;</td><td align='left'>2.20</td><td align='left'>easy to assemble, potential for lower quality consensus</td></tr>
<tr><td>unique             </td><td>71813</td><td>64.82</td><td align='right'>6243.33</td><td>&plusmn;</td><td align='left'>4389.77</td><td align='right'>27.32</td><td>&plusmn;</td><td align='left'>8.13</td><td align='left'>easy to assemble, perfect, yay</td></tr>
<tr><td>repeat-cont        </td><td>17131</td><td>15.46</td><td align='right'>6173.73</td><td>&plusmn;</td><td align='left'>4474.37</td><td align='right'>263.70</td><td>&plusmn;</td><td align='left'>113.55</td><td align='left'>potential for consensus errors, no impact on assembly</td></tr>
<tr><td>repeat-dove          </td><td>103</td><td>0.09</td><td align='right'>20980.99</td><td>&plusmn;</td><td align='left'>5787.85</td><td align='right'>255.55</td><td>&plusmn;</td><td align='left'>105.58</td><td align='left'>hard to assemble, likely won't assemble correctly or even at all</td></tr>
<tr><td>span-repeat        </td><td>11049</td><td>9.97</td><td align='right'>8322.26</td><td>&plusmn;</td><td align='left'>5383.50</td><td align='right'>1766.60</td><td>&plusmn;</td><td align='left'>2251.16</td><td align='left'>read spans a large repeat, usually easy to assemble</td></tr>
<tr><td>uniq-repeat-cont    </td><td>8319</td><td>7.51</td><td align='right'>6848.00</td><td>&plusmn;</td><td align-'left'>4365.35</td><td></td><td></td><td></td><td align='left'>should be uniquely placed, low potential for consensus errors, no impact on assembly</td></tr>
<tr><td>uniq-repeat-dove    </td><td>1293</td><td>1.17</td><td align='right'>13030.59</td><td>&plusmn;</td><td align-'left'>5220.67</td><td></td><td></td><td></td><td align='left'>will end contigs, potential to misassemble</td></tr>
<tr><td>uniq-anchor          </td><td>690</td><td>0.62</td><td align='right'>10541.48</td><td>&plusmn;</td><td align='left'>4118.73</td><td align='right'>3321.16</td><td>&plusmn;</td><td align='left'>3017.42</td><td align='left'>repeat read, with unique section, probable bad read</td></tr>
</table>
<h2>Overlap Error Adjustment</h2>

<h2>Unitigs</h2>

<h3>Overlaps</h3>

Loaded all overlaps for 110785 reads.<br>
Loaded 12501316 overlaps in total.<br>
<h3>Edges</h3>

Constructing unitigs using overlaps of at most this fraction error:<br>
0.00380000 +- 0.00270000 = 0.02781812 = 2.781812% (median absolute deviation)<br>
0.00703128 +- 0.00855913 = 0.05838605 = 5.838605% (standard deviation)<br>
<br>
INITIAL EDGES<br>
102303 reads are contained.<br>
65 reads are singleton.<br>
245 reads are spur (197 have a mutual best edge).<br>
8172 reads form the backbone (52 have no mutual best edges; 375 have one; 7745 have both).<br>
<br>
FILTERING<br>
0 reads have a suspicious overlap pattern.<br>
4978 had one high error rate edge filtered; 1264 had both.<br>
346 had one size incompatible edge filtered; 57 had both.<br>
<br>
FINAL EDGES<br>
101731 reads are contained.<br>
505 reads are singleton.<br>
540 reads are spur (109 have a mutual best edge).<br>
8009 reads form the backbone (339 have no mutual best edges; 177 have one; 7493 have both).<br>
<h3>Initial Tig Sizes</h3>
<h3>Final Tig Sizes</h3>
<h2>Consensus</h2>

<h2>Final Outputs</h2>


<script type='text/javascript'>
var toggleTable = function() {
  var table = this.closest('table');
  var elts  = table.querySelectorAll('.details');

  for (var i=0; i<elts.length; i++) {
    if (!elts[i].enabled) {
      elts[i].enabled = true;
      elts[i].style.visibility = 'visible';
    } else {
      elts[i].enabled = false;
      elts[i].style.visibility = 'collapse';
    }
  }
}
document.getElementById('gkpload0').onclick = toggleTable;
document.getElementById('gkpload0').style   = 'cursor: pointer;';
</script>

</body>

</html>
